tsl/src/compression/create.c

```diff
@@ -1244,74 +1250,114 @@ compression_setting_orderby_get_default(Hypertable *ht, ArrayType *segmentby)
 	{
 		settings->fd.segmentby =
 			ts_compress_hypertable_parse_segment_by(with_clause_options[AlterTableFlagSegmentBy],
 													ht);
 	}
-	else if (!settings->fd.segmentby && !settings->fd.orderby &&
-			 with_clause_options[AlterTableFlagOrderBy].is_default)
+
+	if (!with_clause_options[AlterTableFlagOrderBy].is_default)
 	{
-		settings->fd.segmentby = compression_setting_segmentby_get_default(ht);
+		OrderBySettings obs =
+			ts_compress_hypertable_parse_order_by(with_clause_options[AlterTableFlagOrderBy], ht);
+		obs = add_time_to_order_by_if_not_included(obs, settings->fd.segmentby, ht);
+		settings->fd.orderby = obs.orderby;
+		settings->fd.orderby_desc = obs.orderby_desc;
+		settings->fd.orderby_nullsfirst = obs.orderby_nullsfirst;
 	}
 
-	if (!with_clause_options[AlterTableFlagOrderBy].is_default || !settings->fd.orderby)
+	/* update manual settings */
+	ts_compression_settings_update(settings);
+}
+
+static void
+compression_settings_set_manually_for_create(Hypertable *ht, CompressionSettings *settings,
+											 WithClauseResult *with_clause_options)
+{
+	if (with_clause_options[CreateTableFlagSegmentBy].is_default &&
+		with_clause_options[CreateTableFlagOrderBy].is_default)
+		return;
+
+	if (!with_clause_options[CreateTableFlagSegmentBy].is_default)
 	{
-		OrderBySettings obs;
-		if (with_clause_options[AlterTableFlagOrderBy].is_default)
-		{
-			obs = compression_setting_orderby_get_default(ht, settings->fd.segmentby);
-		}
-		else
-		{
-			obs = ts_compress_hypertable_parse_order_by(with_clause_options[AlterTableFlagOrderBy],
-														ht);
-			obs = add_time_to_order_by_if_not_included(obs, settings->fd.segmentby, ht);
-		}
+		settings->fd.segmentby =
+			ts_compress_hypertable_parse_segment_by(with_clause_options[CreateTableFlagSegmentBy],
+													ht);
+	}
+
+	if (!with_clause_options[CreateTableFlagOrderBy].is_default)
+	{
+		OrderBySettings obs =
+			ts_compress_hypertable_parse_order_by(with_clause_options[CreateTableFlagOrderBy], ht);
+		obs = add_time_to_order_by_if_not_included(obs, settings->fd.segmentby, ht);
 		settings->fd.orderby = obs.orderby;
 		settings->fd.orderby_desc = obs.orderby_desc;
 		settings->fd.orderby_nullsfirst = obs.orderby_nullsfirst;
 	}
 
+	/* update manual settings */
 	ts_compression_settings_update(settings);
 }
 
+static void
+compression_settings_update(Hypertable *ht, CompressionSettings *settings,
+							WithClauseResult *with_clause_options, const bool allow_defaults)
+{
+	/* orderby arrays should always be in sync either all NULL or none */
+	Assert(
+		(settings->fd.orderby && settings->fd.orderby_desc && settings->fd.orderby_nullsfirst) ||
+		(!settings->fd.orderby && !settings->fd.orderby_desc && !settings->fd.orderby_nullsfirst));
+
+	compression_settings_set_manually_for_alter(ht, settings, with_clause_options);
+
+	if (allow_defaults)
+		compression_settings_set_defaults(ht, settings, with_clause_options);
```

Comment:
So this part writes the manual settings into the catalog, and then also applies default settings but only in memory? I think this is a pretty confusing interface for a function, can we straighten it somehow, e.g. make the function only modify the in-memory representation, and require the caller to write it into the catalog?